monai/apps/deepgrow/handler.py

```diff
@@ -0,0 +1,287 @@
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import statistics
+
+import numpy as np
+import torch
+import torch.distributed
+
+from monai.engines.workflow import Engine, Events
+from monai.handlers.tensorboard_handlers import SummaryWriter
+from monai.metrics import compute_meandice
+from monai.transforms import rescale_array
+from monai.utils import optional_import
+from monai.visualize import plot_2d_or_3d_image
+
+nib, _ = optional_import("nibabel")
+torchvision, _ = optional_import("torchvision")
+make_grid, _ = optional_import("torchvision.utils", name="make_grid")
+Image, _ = optional_import("PIL.Image")
+ImageDraw, _ = optional_import("PIL.ImageDraw")
+
+
+# TODO:: Unit Test
+
+
+class MeanDice:
+    def __init__(self):
+        self.data = []
+
+    def reset(self):
+        self.data = []
+
+    def update(self, y_pred, y, batched=True):
+        if not batched:
+            y_pred = y_pred[None]
+            y = y[None]
+        score = compute_meandice(y_pred=y_pred, y=y, include_background=False).mean()
+        self.data.append(score.item())
+
+    def mean(self):
+        return statistics.mean(self.data)
+
+    def stdev(self):
+        return statistics.stdev(self.data) if len(self.data) > 1 else 0
+
+
+class DeepgrowStatsHandler:
+    def __init__(
+        self,
+        summary_writer=None,
+        interval=1,
+        log_dir="./runs",
+        tag_name="val_dice",
+        compute_metric=True,
+        images=True,
+        image_interval=1,
+        max_channels=1,
+        max_frames=64,
+        add_scalar=True,
+        merge_scalar=False,
+        fold_size=0,
+    ):
+        self.writer = SummaryWriter(log_dir=log_dir) if summary_writer is None else summary_writer
+        self.interval = interval
+        self.tag_name = tag_name
+        self.compute_metric = compute_metric
+        self.images = images
+        self.image_interval = image_interval
+        self.max_channels = max_channels
+        self.max_frames = max_frames
+        self.add_scalar = add_scalar
+        self.merge_scalar = merge_scalar
+        self.fold_size = fold_size
+        self.logger = logging.getLogger(__name__)
+
+        if torch.distributed.is_initialized():
+            self.tag_name = "{}-r{}".format(self.tag_name, torch.distributed.get_rank())
+
+        self.plot_data = {}
+        self.metric_data = {}
+
+    def attach(self, engine: Engine) -> None:
+        engine.add_event_handler(Events.ITERATION_COMPLETED(every=self.interval), self, "iteration")
+        engine.add_event_handler(Events.EPOCH_COMPLETED(every=1), self, "epoch")
```

Comment:
do we need some checks here? like
`if not engine.has_event_handler(self, Events.ITERATION_COMPLETED)`

Reply:
removed this code.. not need as part of core functionality for deepgrow